File: app/src/main/java/ch/bailu/aat/dispatcher/EditorSource.java
package ch.bailu.aat.dispatcher;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import ch.bailu.aat.gpx.GpxInformation;
import ch.bailu.aat.services.ServiceContext;
import ch.bailu.aat.services.editor.EditorHelper;
import ch.bailu.aat.services.editor.EditorInterface;
import ch.bailu.aat.util.AppBroadcaster;
import ch.bailu.aat.util.AppIntent;
import ch.bailu.util_java.foc.Foc;

public class EditorSource extends ContentSource implements  EditorSourceInterface{

    private final ServiceContext scontext;
    private final EditorHelper edit;


    private final BroadcastReceiver onFileEdited = new BroadcastReceiver () {
        @Override
        public void onReceive(Context context, Intent intent) {
            if (AppIntent.hasFile(intent, edit.getVID())) {
                requestUpdate();
            }
        }
    };

    
    public EditorSource (ServiceContext sc) {
        scontext = sc;
        edit = new EditorHelper(sc);
    }



    @Override
    public void requestUpdate() {
        sendUpdate(edit.getInfoID(), edit.getInformation());
    }




    @Override
    public void onPause() {
        scontext.getContext().unregisterReceiver(onFileEdited);
        edit.onPause();
    }


    @Override
    public void onResume() {
        AppBroadcaster.register(scontext.getContext(), onFileEdited, AppBroadcaster.FILE_CHANGED_INCACHE);
        edit.onResume();
    }

    @Override
    public boolean isEditing() {
        return true;
    }

    @Override
    public EditorInterface getEditor() {
        return edit.getEditor();
    }

    @Override
    public int getIID() {
        return edit.getInfoID();
    }

    @Override
    public GpxInformation getInfo() {
        return edit.getInformation();
    }

    @Override
    public Foc getFile() {
        return edit.getFile();
    }

    @Override
    public void edit() {

    }

    public int getInfoID() {
        return edit.getInfoID();
    }

    public void edit(Foc file) {
        edit.edit(file);
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/dispatcher/source/EditorSource.kt
package ch.bailu.aat_lib.dispatcher.source

import ch.bailu.aat_lib.app.AppContext
import ch.bailu.aat_lib.broadcaster.AppBroadcaster
import ch.bailu.aat_lib.broadcaster.BroadcastData
import ch.bailu.aat_lib.broadcaster.BroadcastReceiver
import ch.bailu.aat_lib.dispatcher.EditorSourceInterface
import ch.bailu.aat_lib.dispatcher.SourceInterface
import ch.bailu.aat_lib.dispatcher.TargetInterface
import ch.bailu.aat_lib.dispatcher.usage.UsageTrackerInterface
import ch.bailu.aat_lib.gpx.GpxList
import ch.bailu.aat_lib.gpx.information.GpxFileWrapper
import ch.bailu.aat_lib.gpx.information.GpxInformation
import ch.bailu.aat_lib.service.editor.EditorHelper
import ch.bailu.aat_lib.service.editor.EditorInterface
import ch.bailu.foc.Foc

class EditorSource(private val appContext: AppContext, usageTracker: UsageTrackerInterface) : SourceInterface,
    EditorSourceInterface {

    private var target = TargetInterface.NULL
    private val edit: EditorHelper = EditorHelper(appContext)
    private val onFileEdited = BroadcastReceiver { args ->
        if (BroadcastData.has(args, edit.vID)) {
            requestUpdate()
        }
    }

    override fun setTarget(target: TargetInterface) {
        this.target = target
    }

    fun sendUpdate(iid: Int, info: GpxInformation) {
        target.onContentUpdated(iid, info)
    }

    init {
        usageTracker.observe {
            requestUpdate()
        }
    }
    override fun requestUpdate() {
        sendUpdate(edit.infoID, edit.info)
    }

    override fun onPauseWithService() {
        appContext.broadcaster.unregister(onFileEdited)
        edit.onPause()
    }
    override fun onDestroy() {}

    override fun onResumeWithService() {
        appContext.broadcaster.register(AppBroadcaster.FILE_CHANGED_INCACHE, onFileEdited)
        edit.onResume()
    }

    override val isEditing = true

    override val editor: EditorInterface
        get() =  edit.editor

    override fun getIID(): Int {
        return edit.infoID
    }

    override fun getInfo(): GpxInformation {
        return edit.info
    }

    override val file: Foc
        get() = edit.file

    override fun edit() {}

    fun edit(file: Foc) {
        sendUpdate(edit.infoID, GpxFileWrapper(file, GpxList.NULL_ROUTE))
        edit.edit(file)
        requestUpdate()
    }
}
